File: contests/lc/1590.MakeSumDivisiblebyP/solution.go
package main

// minSubarray finds the shortest subarray to remove so the remaining sum is divisible by p.
func minSubarray(nums []int, p int) int {
	// total modulo p; we only need the remainder, not the full sum.
	total := 0
	for _, v := range nums {
		total = (total + v%p) % p
	}
	target := total % p
	if target == 0 {
		return 0
	}

	n := len(nums)
	best := n + 1
	prefix := 0
	pos := map[int]int{0: -1} // remainder -> earliest index

	for i, v := range nums {
		prefix = (prefix + v%p) % p
		need := (prefix - target + p) % p
		if idx, ok := pos[need]; ok && i-idx < best {
			best = i - idx
		}
		pos[prefix] = i
	}

	if best < n {
		return best
	}
	return -1
}

func minSubarray2(nums []int, p int) int {
	lastIdx := make(map[int]int, len(nums))
	lastIdx[0] = -1

	sum := 0
	pref := 0

	for _, v := range nums {
		sum = (sum + v) % p
	}

	ans := len(nums)
	for i := 0; i <= len(nums); i++ {
		suf := (sum - pref + p) % p

		need := (p - suf) % p
		if last, ok := lastIdx[need]; ok {
			ans = min(ans, i-last-1)
		}

		if i < len(nums) {
			pref = (pref + nums[i]) % p
			lastIdx[pref] = i
		}
	}

	if ans == len(nums) {
		return -1
	}

	return ans
}

File: contests/lc/1590.MakeSumDivisiblebyP/solution.py
from typing import List


class Solution:
    def minSubarray(self, nums: List[int], p: int) -> int:
        target = sum(nums) % p
        if target == 0:
            return 0

        best = len(nums) + 1
        prefix = 0
        seen = {0: -1}  # remainder -> earliest index

        for i, num in enumerate(nums):
            prefix = (prefix + num % p) % p
            need = (prefix - target) % p
            if need in seen:
                best = min(best, i - seen[need])
            seen[prefix] = i

        return best if best < len(nums) else -1

    def minSubarray2(self, nums: List[int], p: int) -> int:
        if sum(nums)%p==0:
            return 0
        target = sum(nums) % p
        dic,n = {0:-1},len(nums)
        cur,ret = 0,n
        for i,num in enumerate(nums):
            cur = (cur+num)%p
            if dic.get((cur-target)%p) is not None:
                ret = min(ret,i-dic.get((cur-target)%p))
            dic[cur] = i
        return ret if ret < n else -1

__import__("atexit").register(lambda: open("display_runtime.txt", "w").write("0"))